util/push/push.go

```diff
@@ -234,3 +238,47 @@ func childrenHandler(provider content.Provider) images.HandlerFunc {
 		return descs, nil
 	}
 }
+
+// updateDistributionSourceHandler will update distribution source label after
+// pushing layer successfully.
+//
+// FIXME(fuweid): There is race condition for current design of distribution
+// source label if there are pull/push jobs consuming same layer.
+func updateDistributionSourceHandler(cs content.Store, pushF images.HandlerFunc, ref string) (images.HandlerFunc, error) {
+	updateF, err := docker.AppendDistributionSourceLabel(cs, ref)
+	if err != nil {
+		return nil, err
+	}
+
+	return images.HandlerFunc(func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		var islayer bool
+		var cmd string
+
+		switch desc.MediaType {
+		case images.MediaTypeDockerSchema2Layer, images.MediaTypeDockerSchema2LayerGzip,
+			ocispec.MediaTypeImageLayer, ocispec.MediaTypeImageLayerGzip:
+
+			islayer = true
+			cmd = fmt.Sprintf("layer %v", desc.Digest)
+		case images.MediaTypeDockerSchema2Config, ocispec.MediaTypeImageConfig:
+			cmd = fmt.Sprintf("config %v", desc.Digest)
+		default:
+			cmd = fmt.Sprintf("unknown %v", desc.Digest)
+		}
+
+		showDone := oneOffProgress(ctx, cmd)
```

Comment:
Isn't this progress output repeated now? There is already progress around `Dispatch`

Reply:
nope. only `pushing layers` right now. just add layer pushing status here. It is reasonable? I am ok to remove it if there is overhead. 😂